pkg/scheduler/framework/interface.go

```diff
@@ -50,9 +50,75 @@ type NodeScore struct {
 	Score int64
 }
 
-// NodeToStatusMap contains the statuses of the Nodes where the incoming Pod was not schedulable.
-// A PostFilter plugin that uses this map should interpret absent Nodes as UnschedulableAndUnresolvable.
-type NodeToStatusMap map[string]*Status
+// NodeToStatusReader is a read-only interface of NodeToStatus passed to each PostFilter plugin.
+type NodeToStatusReader interface {
+	Get(nodeName string) *Status
+	Len() int
+	AbsentNodesStatus() *Status
+	ForEach(fn func(nodeName string, status *Status))
```

Comment:
In addition to ForEachStatus, for performant `nodesWherePreemptionMightHelp`, we may also want to have `Filter`

```go
type NodeToStatusReader interface {
...
	// O(n) where n = len(m.nodeToStatus)
	// which is more performant than iterating all nodes with Get().
	Filter(nodenames []string, fn func(status *Status) bool) []string
}
```

Reply:
Okay, understood (along with [this one](https://github.com/kubernetes/kubernetes/pull/126022#discussion_r1682894392) too). 
Then, performance pov aside, I'd still love to see if we could handle absent node status implicitly under `NodeToStatusReader` and eliminate `AbsentNodesStatus`.

As I wrote in [another comment](https://github.com/kubernetes/kubernetes/pull/126022#discussion_r1681032547), the current `NodeToStatusReader` requires people to be aware of `AbsentNodesStatus()` and handle it separately, which implies the same problem you pointed out in my previous proposal.

> Especially, absent status won't be accidentally ignored in the future, which could happen if handled separately.

Noticing the need to use a new parameter for a correct preemption vs the need to use `AbsentNodesStatus()` for a correct preemption, both include the same cognitive cost. 
(I'm not trying to stick to and revive my previous proposal here, it's just that I'd like to explore the possibility to eliminate this cognitive cost for users by adjusting an interface.)

Reply:
You're right, removing `AbsentNodeStatus()` from the interface can reduce possible risks. 

I added a new method `NodesForStatusCode` which behaves like `nodesWherePreemptionMightHelp`, so it is returning all nodes that have a given status code, both explicitly and implicitly set. Now the `NodeToStatusReader` contains only `Get` and `NodesForStatusCode` methods.

I also renamed `ForEach` to `ForEachExplicitNode` to emphasize it's called only for the nodes explicitly set in the map. It is removed from the `NodeToStatusReader` so the risks of misusing it are lower.

Thanks for the pseudocode as well. I think, that keeping `NodeInfo` in `NodeToStatus` might overcomplicate the usage of the map (`Set` method), because in most cases we don't have corresponding `NodeInfo` available, so would need to get it from the snapshot then. Overall the `Filter` method structure is similar to the `NodesForStatusCode`.